File: src/epub_optimizer/calibre.py
"""Select local Calibre books and update their formats through calibredb."""

import json
import os
import shutil
import subprocess
import sys
import tempfile
from dataclasses import dataclass
from pathlib import Path

from epub_optimizer.core import optimize_epub, preview_epub_changes


@dataclass(frozen=True)
class Book:
    id: int
    title: str
    authors: str
    formats: dict[str, Path]
    directory: Path


def run_calibredb(arguments: list[str], library: Path | None) -> str:
    command = ["calibredb", *arguments]
    if library is not None:
        command.extend(["--with-library", str(library)])
    try:
        result = subprocess.run(command, capture_output=True, text=True, check=False)
    except FileNotFoundError as exc:
        raise RuntimeError("calibredb was not found. Install Calibre and put it on PATH.") from exc
    if result.returncode:
        detail = result.stderr.strip() or result.stdout.strip() or f"exit {result.returncode}"
        raise RuntimeError(f"calibredb {arguments[0]} failed: {detail}")
    return result.stdout


def _pick(labels: list[str], prompt: str) -> int:
    if len(labels) == 1:
        return 0
    rows = [f"{index}\t{' '.join(label.split())}" for index, label in enumerate(labels)]
    # User defaults such as --multi, --filter or --expect change the output protocol.
    env = dict(os.environ)
    env.pop("FZF_DEFAULT_OPTS", None)
    env.pop("FZF_DEFAULT_OPTS_FILE", None)
    try:
        result = subprocess.run(
            ["fzf", "--no-multi", "--delimiter=\t", "--with-nth=2..", f"--prompt={prompt}"],
            input="\n".join(rows) + "\n",
            stdout=subprocess.PIPE,
            text=True,
            env=env,
            check=False,
        )
    except FileNotFoundError as exc:
        raise RuntimeError("Multiple matches found. Install fzf to choose one.") from exc
    if result.returncode in {1, 130}:
        raise KeyboardInterrupt
    if result.returncode:
        raise RuntimeError(f"fzf failed with exit code {result.returncode}.")
    selected = result.stdout.rstrip("\n")
    if selected not in rows:
        raise RuntimeError("fzf returned an invalid selection.")
    return rows.index(selected)


def _list_entries(query: str, library: str | None) -> tuple[list[dict], Path | None]:
    root = None
    if library:
        root = Path(library).expanduser().resolve()
        if not (root / "metadata.db").is_file():
            raise ValueError(f"Not a local Calibre library, metadata.db is missing: {root}")
    raw = run_calibredb(
        ["list", "--for-machine", "--fields", "title,authors,formats,cover", "--search", query],
        root,
    )
    try:
        entries = json.loads(raw)
        if not isinstance(entries, list):
            raise ValueError("expected a list of books")
        for row in entries:
            int(row["id"])
            row["title"]
    except (ValueError, KeyError, TypeError) as exc:
        raise RuntimeError("calibredb returned an invalid book list.") from exc
    return entries, root


def _entry_book(entry: dict, root: Path | None) -> tuple[Book, Path]:
    formats = {Path(path).suffix[1:].upper(): Path(path) for path in entry.get("formats", [])}
    paths = list(formats.values())
    if entry.get("cover"):
        paths.append(Path(entry["cover"]))
    if not paths:
        raise ValueError(f"No local files found for book {entry['id']}: {entry['title']}")
    directory = paths[0].parent
    # Pin the default library for subsequent calls, even if the GUI preference changes.
    if root is None:
        root = next((p for p in directory.parents if (p / "metadata.db").is_file()), None)
        if root is None:
            raise ValueError(f"Could not locate the Calibre library containing {directory}")
    book = Book(int(entry["id"]), entry["title"], entry.get("authors", ""), formats, directory)
    return book, root


def select_epub(query: str, library: str | None) -> tuple[Book, Path, Path]:
    entries, root = _list_entries(query, library)
    if not entries:
        raise ValueError(f"No Calibre books match: {query}")
    labels = [f"{row['id']} | {row['title']} | {row.get('authors', '')}" for row in entries]
    book, root = _entry_book(entries[_pick(labels, "Book> ")], root)
    candidates = sorted(
        (
            path for path in book.directory.iterdir()
            if path.is_file() and path.suffix.lower() == ".epub"
            and not path.name.lower().endswith(".original.epub")
        ),
        key=lambda path: path.name.casefold(),
    )
    if not candidates:
        raise ValueError(f"No EPUB files found for book {book.id}: {book.title}")
    source = candidates[_pick([path.name for path in candidates], "EPUB> ")]
    return book, source, root


def normalize_calibre(
    query: str,
    library: str | None = None,
    *,
    all_books: bool = False,
    dry_run: bool = False,
    preserve_publisher_css: bool = False,
    verbose: bool = False,
) -> int:
    if all_books:
        return _normalize_all(
            library, dry_run=dry_run,
            preserve_publisher_css=preserve_publisher_css, verbose=verbose,
        )
    book, source, root = select_epub(query, library)
    print(f"Selected {book.id}: {book.title} | {book.authors}")
    _normalize_book(
        book, source, root, dry_run=dry_run,
        preserve_publisher_css=preserve_publisher_css, verbose=verbose,
    )
    return 0


def _normalize_all(
    library: str | None,
    *,
    dry_run: bool,
    preserve_publisher_css: bool,
    verbose: bool,
) -> int:
    # Query the database, not filenames: ORIGINAL_EPUB is a distinct registered format.
    entries, root = _list_entries('formats:"=EPUB" and not formats:"=ORIGINAL_EPUB"', library)
    if not entries:
        print("No books with EPUB and without ORIGINAL_EPUB need processing.")
        return 0
    failed: list[int] = []
    for index, entry in enumerate(entries, 1):
        print(f"[{index}/{len(entries)}] {entry['title']} | {entry.get('authors', '')}", flush=True)
        try:
            book, root = _entry_book(entry, root)
            source = book.formats.get("EPUB")
            if source is None:
                raise ValueError("Registered EPUB file is missing.")
            _normalize_book(
                book, source, root, dry_run=dry_run,
                preserve_publisher_css=preserve_publisher_css, verbose=verbose,
            )
        except Exception as exc:
            failed.append(int(entry["id"]))
            print(f"Book {entry['id']} ({entry['title']}): error: {exc}", file=sys.stderr)
    action = "previewed" if dry_run else "processed"
    print(f"Batch complete: {len(entries) - len(failed)} {action}, {len(failed)} failed.")
    if failed:
        print(f"Failed book IDs: {', '.join(map(str, failed))}", file=sys.stderr)
    return int(bool(failed))


def _normalize_book(
    book: Book,
    source: Path,
    root: Path,
    *,
    dry_run: bool,
    preserve_publisher_css: bool,
    verbose: bool,
) -> None:
    print(f"EPUB: {source}")
    current = book.formats.get("EPUB", source)
    original = book.formats.get("ORIGINAL_EPUB")
    if dry_run:
        preview = preview_epub_changes(source, preserve_publisher_css=preserve_publisher_css)
        print(
            f"[dry run] {preview.content_documents} document(s), "
            f"{preview.stylesheets_and_fonts} stylesheet/font entry(s) to remove, "
            f"{preview.images_preserved} image(s) preserved"
        )
        if original:
            print(f"Would keep ORIGINAL_EPUB: {original}")
        else:
            print(f"Would back up as ORIGINAL_EPUB: {current}")
        print(f"Would replace EPUB on Calibre book {book.id} using add_format.")
        for warning in preview.warnings + preview.image_diagnostics:
            print(f"warning: {warning}", file=sys.stderr)
        return

    def progress(message: str) -> None:
        print(message, file=sys.stderr)

    with tempfile.TemporaryDirectory(prefix="epub-normalize-calibre-") as tmp:
        work = Path(tmp)
        backup = work / current.with_suffix(".original_epub").name
        if original is None:
            shutil.copyfile(current, backup)
        result = optimize_epub(
            source,
            work,
            output_filename=current.name,
            preserve_publisher_css=preserve_publisher_css,
            progress=progress if verbose else None,
        )
        # Only change the library after normalization succeeds. Register the backup first.
        if original is None:
            run_calibredb(["add_format", str(book.id), str(backup), "--dont-replace"], root)
            print("Saved ORIGINAL_EPUB backup in Calibre.")
        else:
            print(f"Kept existing ORIGINAL_EPUB: {original}")
        try:
            run_calibredb(["add_format", str(book.id), str(result.output_path)], root)
        except RuntimeError as exc:
            raise RuntimeError(
                f"{exc}\nThe ORIGINAL_EPUB backup is retained in Calibre."
            ) from exc
    print(f"Updated Calibre book {book.id}. Formats: EPUB, ORIGINAL_EPUB.")
    print(f"  EPUBCheck: {result.validation_outcome}")
    for warning in result.warnings + result.image_diagnostics:
        print(f"warning: {warning}", file=sys.stderr)

File: src/calibre.rs
use crate::{
    archive,
    cli::{self, Args},
    core::{self, Options},
    process::{self, Interrupted},
};
use anyhow::{Result, anyhow, bail};
use serde_json::Value;
use std::{
    collections::BTreeMap,
    fs,
    io::Write,
    path::{Path, PathBuf},
    process::Command,
};
use unicode_casefold::UnicodeCaseFold;

pub struct Book {
    pub id: i64,
    pub title: String,
    pub authors: String,
    pub formats: BTreeMap<String, PathBuf>,
    pub directory: PathBuf,
}
pub fn run_calibredb(arguments: &[String], library: Option<&Path>) -> Result<String> {
    let mut command = Command::new("calibredb");
    command.args(arguments);
    if let Some(library) = library {
        command.arg("--with-library").arg(library);
    }
    let result = process::capture(&mut command, None, false, None).map_err(|e| {
        if e.downcast_ref::<std::io::Error>()
            .is_some_and(|e| e.kind() == std::io::ErrorKind::NotFound)
        {
            anyhow!("calibredb was not found. Install Calibre and put it on PATH.")
        } else {
            e
        }
    })?;
    if !result.status.success() {
        let fallback = format!("exit {}", result.status.code().unwrap_or(-1));
        let detail = if !result.stderr.trim().is_empty() {
            result.stderr.trim()
        } else if !result.stdout.trim().is_empty() {
            result.stdout.trim()
        } else {
            &fallback
        };
        bail!("calibredb {} failed: {detail}", arguments[0]);
    }
    Ok(result.stdout)
}
pub fn pick(labels: &[String], prompt: &str) -> Result<usize> {
    if labels.len() == 1 {
        return Ok(0);
    }
    let rows: Vec<_> = labels
        .iter()
        .enumerate()
        .map(|(i, l)| {
            format!(
                "{i}\t{}",
                l.split_whitespace().collect::<Vec<_>>().join(" ")
            )
        })
        .collect();
    let mut command = Command::new("fzf");
    command
        .args([
            "--no-multi",
            "--delimiter=\t",
            "--with-nth=2..",
            &format!("--prompt={prompt}"),
        ])
        .env_remove("FZF_DEFAULT_OPTS")
        .env_remove("FZF_DEFAULT_OPTS_FILE");
    let output = process::capture(
        &mut command,
        Some(&format!("{}\n", rows.join("\n"))),
        true,
        None,
    )
    .map_err(|e| {
        if e.downcast_ref::<std::io::Error>()
            .is_some_and(|e| e.kind() == std::io::ErrorKind::NotFound)
        {
            anyhow!("Multiple matches found. Install fzf to choose one.")
        } else {
            e
        }
    })?;
    if matches!(output.status.code(), Some(1 | 130)) {
        return Err(Interrupted.into());
    }
    if !output.status.success() {
        bail!(
            "fzf failed with exit code {}.",
            output.status.code().unwrap_or(-1)
        );
    }
    rows.iter()
        .position(|r| r == output.stdout.trim_end_matches('\n'))
        .ok_or_else(|| anyhow!("fzf returned an invalid selection."))
}
fn entry_id(entry: &Value) -> Option<i64> {
    entry.get("id").and_then(|v| {
        v.as_i64()
            .or_else(|| v.as_str().and_then(|s| s.parse().ok()))
    })
}
fn title(entry: &Value) -> String {
    entry
        .get("title")
        .and_then(Value::as_str)
        .unwrap_or("")
        .into()
}
fn authors(entry: &Value) -> String {
    entry
        .get("authors")
        .and_then(Value::as_str)
        .unwrap_or("")
        .into()
}
fn list_entries(query: &str, library: Option<&str>) -> Result<(Vec<Value>, Option<PathBuf>)> {
    let root = if let Some(library) = library.filter(|s| !s.is_empty()) {
        let expanded = if library == "~" || library.starts_with("~/") {
            std::env::var("HOME")
                .map(|home| format!("{home}{}", &library[1..]))
                .unwrap_or_else(|_| library.into())
        } else {
            library.into()
        };
        let root = cli::resolve(Path::new(&expanded))?;
        if !root.join("metadata.db").is_file() {
            bail!(
                "Not a local Calibre library, metadata.db is missing: {}",
                root.display()
            );
        }
        Some(root)
    } else {
        None
    };
    let raw = run_calibredb(
        &[
            "list",
            "--for-machine",
            "--fields",
            "title,authors,formats,cover",
            "--search",
            query,
        ]
        .map(str::to_string),
        root.as_deref(),
    )?;
    let entries: Vec<Value> = serde_json::from_str(&raw)
        .map_err(|_| anyhow!("calibredb returned an invalid book list."))?;
    if entries
        .iter()
        .any(|e| entry_id(e).is_none() || e.get("title").is_none())
    {
        bail!("calibredb returned an invalid book list.");
    }
    Ok((entries, root))
}
fn entry_book(entry: &Value, root: Option<&Path>) -> Result<(Book, PathBuf)> {
    let paths: Vec<_> = entry
        .get("formats")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .filter_map(Value::as_str)
        .map(PathBuf::from)
        .collect();
    let formats = paths
        .iter()
        .map(|p| (archive::extension(p).to_uppercase(), p.clone()))
        .collect();
    let cover = entry
        .get("cover")
        .and_then(Value::as_str)
        .filter(|s| !s.is_empty())
        .map(PathBuf::from);
    let first = paths.first().or(cover.as_ref()).ok_or_else(|| {
        anyhow!(
            "No local files found for book {}: {}",
            entry_id(entry).unwrap(),
            title(entry)
        )
    })?;
    let directory = first.parent().unwrap_or(Path::new(".")).to_path_buf();
    let root = root
        .map(Path::to_path_buf)
        .or_else(|| {
            directory
                .ancestors()
                .skip(1)
                .find(|p| p.join("metadata.db").is_file())
                .map(Path::to_path_buf)
        })
        .ok_or_else(|| {
            anyhow!(
                "Could not locate the Calibre library containing {}",
                directory.display()
            )
        })?;
    Ok((
        Book {
            id: entry_id(entry).unwrap(),
            title: title(entry),
            authors: authors(entry),
            formats,
            directory,
        },
        root,
    ))
}
pub fn normalize(query: &str, args: &Args) -> Result<i32> {
    if args.all {
        return normalize_all(args);
    }
    let (entries, root) = list_entries(query, args.with_library.as_deref())?;
    if entries.is_empty() {
        bail!("No Calibre books match: {query}");
    }
    let labels: Vec<_> = entries
        .iter()
        .map(|e| format!("{} | {} | {}", entry_id(e).unwrap(), title(e), authors(e)))
        .collect();
    let (book, root) = entry_book(&entries[pick(&labels, "Book> ")?], root.as_deref())?;
    let mut candidates: Vec<_> = fs::read_dir(&book.directory)?
        .collect::<std::io::Result<Vec<_>>>()?
        .into_iter()
        .map(|e| e.path())
        .filter(|p| {
            p.is_file()
                && archive::extension(p) == "epub"
                && !p
                    .file_name()
                    .unwrap_or_default()
                    .to_string_lossy()
                    .to_lowercase()
                    .ends_with(".original.epub")
        })
        .collect();
    candidates.sort_by_cached_key(|p| {
        p.file_name()
            .unwrap_or_default()
            .to_string_lossy()
            .case_fold()
            .collect::<String>()
    });
    if candidates.is_empty() {
        bail!("No EPUB files found for book {}: {}", book.id, book.title);
    }
    let labels = candidates
        .iter()
        .map(|p| {
            p.file_name()
                .unwrap_or_default()
                .to_string_lossy()
                .into_owned()
        })
        .collect::<Vec<_>>();
    let source = &candidates[pick(&labels, "EPUB> ")?];
    println!("Selected {}: {} | {}", book.id, book.title, book.authors);
    normalize_book(&book, source, &root, args)?;
    Ok(0)
}
fn normalize_all(args: &Args) -> Result<i32> {
    let (entries, mut root) = list_entries(
        "formats:\"=EPUB\" and not formats:\"=ORIGINAL_EPUB\"",
        args.with_library.as_deref(),
    )?;
    if entries.is_empty() {
        println!("No books with EPUB and without ORIGINAL_EPUB need processing.");
        return Ok(0);
    }
    let mut failed = Vec::new();
    for (index, entry) in entries.iter().enumerate() {
        process::check_interrupt()?;
        println!(
            "[{}/{}] {} | {}",
            index + 1,
            entries.len(),
            title(entry),
            authors(entry)
        );
        std::io::stdout().flush()?;
        let result = (|| {
            let (book, library) = entry_book(entry, root.as_deref())?;
            root = Some(library.clone());
            let source = book
                .formats
                .get("EPUB")
                .ok_or_else(|| anyhow!("Registered EPUB file is missing."))?;
            normalize_book(&book, source, &library, args)
        })();
        if let Err(error) = result {
            if error.is::<Interrupted>() {
                return Err(error);
            }
            let id = entry_id(entry).unwrap();
            failed.push(id);
            eprintln!("Book {id} ({}): error: {error}", title(entry));
        }
    }
    println!(
        "Batch complete: {} {}, {} failed.",
        entries.len() - failed.len(),
        if args.dry_run {
            "previewed"
        } else {
            "processed"
        },
        failed.len()
    );
    if !failed.is_empty() {
        eprintln!(
            "Failed book IDs: {}",
            failed
                .iter()
                .map(i64::to_string)
                .collect::<Vec<_>>()
                .join(", ")
        );
    }
    Ok(i32::from(!failed.is_empty()))
}
fn normalize_book(book: &Book, source: &Path, root: &Path, args: &Args) -> Result<()> {
    println!("EPUB: {}", source.display());
    let current = book
        .formats
        .get("EPUB")
        .map(PathBuf::as_path)
        .unwrap_or(source);
    let original = book.formats.get("ORIGINAL_EPUB");
    if args.dry_run {
        let preview = core::preview(source, args.preserve_publisher_css)?;
        println!(
            "[dry run] {} document(s), {} stylesheet/font entry(s) to remove, {} image(s) preserved",
            preview.content_documents, preview.stylesheets_and_fonts, preview.images_preserved
        );
        if let Some(original) = original {
            println!("Would keep ORIGINAL_EPUB: {}", original.display());
        } else {
            println!("Would back up as ORIGINAL_EPUB: {}", current.display());
        }
        println!(
            "Would replace EPUB on Calibre book {} using add_format.",
            book.id
        );
        for warning in preview.warnings.iter().chain(&preview.image_diagnostics) {
            eprintln!("warning: {warning}");
        }
        return Ok(());
    }
    let temp = tempfile::Builder::new()
        .prefix("epub-normalize-calibre-")
        .tempdir()?;
    let backup = temp
        .path()
        .join(current.with_extension("original_epub").file_name().unwrap());
    if original.is_none() {
        fs::copy(current, &backup)?;
    }
    let progress = |message: &str| eprintln!("{message}");
    let filename = current.file_name().unwrap_or_default().to_string_lossy();
    let result = core::optimize(
        source,
        temp.path(),
        &Options {
            output_filename: Some(&filename),
            preserve_publisher_css: args.preserve_publisher_css,
            progress: args.verbose.then_some(&progress),
            ..Default::default()
        },
    )?;
    process::check_interrupt()?;
    if let Some(original) = original {
        println!("Kept existing ORIGINAL_EPUB: {}", original.display());
    } else {
        run_calibredb(
            &[
                "add_format".into(),
                book.id.to_string(),
                backup.to_string_lossy().into_owned(),
                "--dont-replace".into(),
            ],
            Some(root),
        )?;
        println!("Saved ORIGINAL_EPUB backup in Calibre.");
    }
    run_calibredb(
        &[
            "add_format".into(),
            book.id.to_string(),
            result.output_path.to_string_lossy().into_owned(),
        ],
        Some(root),
    )
    .map_err(|e| {
        if e.is::<Interrupted>() {
            e
        } else {
            anyhow!("{e}\nThe ORIGINAL_EPUB backup is retained in Calibre.")
        }
    })?;
    println!(
        "Updated Calibre book {}. Formats: EPUB, ORIGINAL_EPUB.",
        book.id
    );
    println!("  EPUBCheck: {}", result.validation_outcome);
    for warning in result.warnings.iter().chain(&result.image_diagnostics) {
        eprintln!("warning: {warning}");
    }
    Ok(())
}
